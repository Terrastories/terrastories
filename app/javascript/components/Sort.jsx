import React, { Component } from 'react';
import Select from 'react-select';
import PropTypes from 'prop-types';

const SORT_CRITERIA = ['most_recent', 'alphabetical', 'reversed_alphabetical'];

class Sort extends Component {
  constructor(props){
    super(props);
    this.state = {
      sortSelectValue: SORT_CRITERIA[0]
    };
  }

  static propTypes = {
    stories: PropTypes.array,
    handleStoriesChanged: PropTypes.func
  };

  static defaultProps = {
    stories: [],
    handleStoriesChanged: () => {}
  };

  componentDidMount() {
    this.handleSort(this.option(this.state.sortSelectValue));
  }

  optionsHash = values => {
    return values.map(value => {
      return this.option(value);
    });
  }

  option = value => {
    return {value: value, label: I18n.t(value)};
  }

  handleSort = (option) => {
    this.setState({
      sortSelectValue: option.value
    });

    let sortedStories = [];

    switch (option.value) {
      case SORT_CRITERIA[0]: {
        // most recent
        sortedStories = this.props.stories.sort((story1, story2) => {
          if (story1.created_at < story2.created_at) {
            return 1;
          }

          if (story1.created_at > story2.created_at) {
            return -1;
          }

          return 0;
        });

        break;
      }
      case SORT_CRITERIA[1]: {
        sortedStories = this.props.stories.sort((story1, story2) => {
          if (story1.title > story2.title) {
            return 1;
          }

          if (story1.title < story2.title) {
            return -1;
          }

          return 0;
        });

        break;
      }
      case SORT_CRITERIA[2]: {
        sortedStories = this.props.stories.sort((story1, story2) => {
          if (story1.title < story2.title) {
            return 1;
          }

          if (story1.title > story2.title) {
            return -1;
          }

          return 0;
        });

        break;
      }
    }

    this.props.handleStoriesChanged(sortedStories);
  }

  render() {
    return (
      <React.Fragment>
        <span className="card--nav-sort">{I18n.t("sort_stories")}: </span>
        <Select
          className="storiesSort"
          classNamePrefix="select"
          name="sort-stories"
          onChange={this.handleSort}
          value={this.option(this.state.sortSelectValue)}
          options={this.optionsHash(SORT_CRITERIA)}
        />
      </React.Fragment>
    );
  }
}

export default Sort;
