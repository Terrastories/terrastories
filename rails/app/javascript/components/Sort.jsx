import React, { Component } from 'react';
import Select from 'react-select';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';

const SORT_CRITERIA = ['most_recent', 'alphabetical', 'reversed_alphabetical'];

class Sort extends Component {
  constructor(props){
    super(props);
    this.state = {
      sortSelectValue: SORT_CRITERIA[0]
    };
  }

  static propTypes = {
    stories: PropTypes.array,
    handleStoriesChanged: PropTypes.func
  };

  static defaultProps = {
    stories: [],
    handleStoriesChanged: () => {}
  };

  componentDidMount() {
    this.handleSort(this.option(this.state.sortSelectValue));
  }

  optionsHash = values => {
    return values.map(value => {
      return this.option(value);
    });
  }

  option = value => {
    return {value: value, label: this.props.t(value)};
  }

  storyPinnedSort = (sortedStories) => (
    sortedStories.sort(story => story.story_pinned ? -1 : 1)
  )

  handleSort = (option) => {
    this.setState({
      sortSelectValue: option.value
    });

    let sortedStories = [];

    switch (option.value) {
      case SORT_CRITERIA[0]: {
        // most recent
        sortedStories = this.props.stories.sort((story1, story2) => {
          if (story1.created_at < story2.created_at) {
            return 1;
          }

          if (story1.created_at > story2.created_at) {
            return -1;
          }

          return 0;
        });

        break;
      }
      case SORT_CRITERIA[1]: {
        sortedStories = this.props.stories.sort((story1, story2) => {
          if (story1.title > story2.title) {
            return 1;
          }

          if (story1.title < story2.title) {
            return -1;
          }

          return 0;
        });

        break;
      }
      case SORT_CRITERIA[2]: {
        sortedStories = this.props.stories.sort((story1, story2) => {
          if (story1.title < story2.title) {
            return 1;
          }

          if (story1.title > story2.title) {
            return -1;
          }

          return 0;
        });

        break;
      }
    }

    sortedStories = this.storyPinnedSort(sortedStories)

    this.props.handleStoriesChanged(sortedStories);
  }

  render() {
    return (
      <React.Fragment>
        <span className="card--nav-sort">{this.props.t("sort_stories")}: </span>
        <Select
          className="storiesSort"
          classNamePrefix="select"
          name="sort-stories"
          onChange={this.handleSort}
          value={this.option(this.state.sortSelectValue)}
          options={this.optionsHash(SORT_CRITERIA)}
        />
      </React.Fragment>
    );
  }
}

export default withTranslation()(Sort);
