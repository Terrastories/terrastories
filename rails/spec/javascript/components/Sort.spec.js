import React from 'react';
import { shallow } from 'enzyme';
import Sort from '../../../app/javascript/components/Sort';

describe('Sort component', () => {
  beforeEach(() => {
    global.I18n = {
      t: jest.fn(),
    };
  });

  const corinneStory = {
    created_at: '2020-10-03T11:13:23.037Z',
    title: "Corinne's testimonial",
  };

  const rudoStory = {
    created_at: '2020-10-01T11:13:23.013Z',
    title: "Rudo's testimonial",
  };

  const sortValues = [
    'sort_stories',
    'most_recent',
    'alphabetical',
    'reversed_alphabetical',
  ];

  it('Displays correctly', () => {
    const wrapper = shallow(<Sort stories={[corinneStory, rudoStory]} />);

    expect(wrapper).toMatchSnapshot();

    sortValues.forEach((sortValue) => {
      expect(global.I18n.t).toHaveBeenCalledWith(sortValue);
    });
  });

  it('Calls handleStoriesChanged on change with expected sort', () => {
    const handleStoriesChanged = jest.fn();

    const props = {
      handleStoriesChanged,
      stories: [corinneStory, rudoStory],
    };
    const wrapper = shallow(<Sort {...props} />);

    // componentDidMount call
    expect(handleStoriesChanged.mock.calls.length).toBe(1);
    expect(handleStoriesChanged.mock.calls[0][0]).toEqual([
      corinneStory,
      rudoStory,
    ]);

    wrapper.find('.storiesSort').simulate('change', { value: 'alphabetical' });

    expect(handleStoriesChanged.mock.calls.length).toBe(2);
    expect(handleStoriesChanged.mock.calls[1][0]).toEqual([
      corinneStory,
      rudoStory,
    ]);

    wrapper
      .find('.storiesSort')
      .simulate('change', { value: 'reversed_alphabetical' });

    expect(handleStoriesChanged.mock.calls.length).toBe(3);
    expect(handleStoriesChanged.mock.calls[2][0]).toEqual([
      rudoStory,
      corinneStory,
    ]);
  });
});
