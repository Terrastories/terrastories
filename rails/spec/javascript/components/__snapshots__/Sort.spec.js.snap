// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort component Displays correctly 1`] = `
<Fragment>
  <span
    className="card--nav-sort"
  >
    : 
  </span>
  <StateManager
    className="storiesSort"
    classNamePrefix="select"
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    name="sort-stories"
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": undefined,
          "value": "most_recent",
        },
        Object {
          "label": undefined,
          "value": "alphabetical",
        },
        Object {
          "label": undefined,
          "value": "reversed_alphabetical",
        },
      ]
    }
    value={
      Object {
        "label": undefined,
        "value": "most_recent",
      }
    }
  />
</Fragment>
`;
